### PR TITLE
Add "preserve" quote-style to mimic Black's skip-string-normalization

### LINKED ISSUE #7525: Formatter: Add --skip-string-normalization option (parity with Black)?
Hey folks. I see that Black added --skip-string-normalization ( https://github.com/psf/black/issues/118#issuecomment-393298377 ) but `ruff format --skip-string-normalization` leads to `error: unexpected argument '--skip-string-normalization' found`. Any chance of adding this feature to ruff format?

Alternatively, preferably, have ruff potentially prefer `''` rather than `""` for strings as an option. That is, still normalize strings, but to `''` if specified. 

I saw two other issues include this text, but neither seemed to be related to this actual functionality. Sorry if it's already addressed.

### PR DESCRIPTION
**Pull request by sciyoshi**
## Summary

Adds a new option to the `quote-style` formatter configuration called "preserve":

* normalizes the prefixes of all strings
* enforces double quotes for triple-quoted strings (docstrings and multiline strings)
* normalizes unnecessary escapes

In other words, using `preserve` leaves the quotes of inline-strings as is, but performs all other normalisation, the same as for `single` or `double`. 

Fixes #7525

## Test Plan

Added tests.  Tested that `quote-style="preserve` without `--preview` fails with an error.

Edited by @MichaReiser: I expanded the summary to be explicit about what `preserve` does and updated the test plan.

### CODE REVIEW

**Review by hussein-awala — COMMENTED** [review]
Nice one! +1

**Comment by MichaReiser** [comment]
Thanks for implementing this new option. It is a good signal on how important this functionality is for the community. 

I haven't looked at the code yet and have some more catch-up work to do before I'll find the time for the review. I'm overall okay with adding this new option but I want to make sure that it removes the need for `--skip-string-normalization` because I want to avoid eventually having to support both.

**Comment by charliermarsh** [comment]
Awesome, thanks @sciyoshi, great to see you back on the repo :) I have basically the same reaction: I support adding this, though it'd be nice if doing so could definitively preclude us from adding `--skip-string-normalization` in the future.

**Comment by sciyoshi** [comment]
Thanks :) I think there's two cases to think about for the `preserve` option and how it compares to `skip-string-normalization`:

1. Whether it should keep existing string prefixes or normalize them 
2. Whether it should format docstrings to use double quotes

Black's behavior with `skip-string-normalization` is to not touch any string prefixes or docstrings. Personally, I think that having ruff normalize string prefixes in `preserve` mode is OK, although docstrings probably should be kept as-is. As it stands, this PR does both (but for #2, only if the docstring is triple-quoted).

**Comment by MichaReiser** [comment]
> Black's behavior with skip-string-normalization is to not touch any string prefixes or docstrings, and also to skip changing any character escapes. Personally, I think that having ruff normalize string prefixes and escapes (1. and 3.) in preserve mode is OK, although docstrings probably should be kept as-is. As it stands, this PR does all three (but for 2., only if the docstring is triple-quoted).

To make sure I understand you correctly. This PR

* normalizes the prefixes of all strings
* enforces double quotes for triple-quoted strings (docstrings and multiline strings)
* normalizes unnecessary escapes

In other words, using `preserve` leaves the quotes of inline-strings as is, but performs all other normalisation as with `single` or `double`.

**Comment by sciyoshi** [comment]
Correct, that's the current behavior of this PR.

**Comment by edreamleo** [comment]
@MichaReiser

The "preserve" quote style looks like a step in the right direction. 

What would this PR do about this?
```python
s = '''
"""
Example docstring
"""
'''
```
Or this?
```python
return('''
"""
Example docstring
"""
'''
)
```
Are you *sure* you know what an "unnecessary" escape is? What about?
```python
pattern = re.compile('''
whatever
''')
```
Or this?
```python
pattern = re.compile(fr"""
whatever
""")
```
Finally, what about backslash-newlines? 
```python
s = textwrap.dedent("""\
unit test data
""")
```
Removing the trailing backslash would break some of Leo's unit tests.

**Comment by zanieb** [comment]
@edreamleo the result for your examples would be

```python
import re
import textwrap

s = '''"""
Example docstring
"""
'''

def foo():
    return('''
    """
    Example docstring
    """
    '''
    )

pattern = re.compile('''
whatever
''')

pattern = re.compile(fr"""
whatever
""")

s = textwrap.dedent("""\
unit test data
""")
```

```diff
--- example.py
+++ example.py
@@ -6,22 +6,29 @@
 """
 '''
 
+
 def foo():
-    return('''
+    return '''
     """
     Example docstring
     """
     '''
-    )
 
-pattern = re.compile('''
+
+pattern = re.compile(
+    """
 whatever
-''')
+"""
+)
 
-pattern = re.compile(fr"""
+pattern = re.compile(
+    rf"""
 whatever
-""")
+"""
+)
 
-s = textwrap.dedent("""\
+s = textwrap.dedent(
+    """\
 unit test data
-""")
+"""
+)
```

This happens to be identical to the formatting _without_ the `preserve` option.

**Comment by edreamleo** [comment]
@zanieb Thanks for your comment!

My only remaining concern is the backslash-newline question. By "whatever" I meant various cases containing escapes that might be considered "unnecessary". Perhaps this PR has no effect on that question.

**Comment by sciyoshi** [comment]
I've added an argument to `ruff format` which will allow setting `--quote-style` via the CLI.

Backslash-newlines will be preserved, as will other escape sequences. From my testing I think it is only unnecessarily-escaped quotes that are changed.

If @MichaReiser and @charliermarsh are OK with it, I think this could probably be merged as-is. Nothing precludes us from changing the behavior of `preserve` to be weaker if there are still concerns with the reformatting of triple quotes & string prefixes - making it stronger on the other hand would be a more of a breaking change. If on the other hand we want `preserve` to match `skip-string-normalization` exactly, I can take a look at changing that behavior as well. We probably do also want a test or two for this option.

File: crates/ruff_python_formatter/src/options.rs [hunk 1]
```diff
@@ -207,35 +207,7 @@ pub enum QuoteStyle {
...
-    type Error = ();
-
-    fn try_from(value: char) -> std::result::Result<Self, Self::Error> {
-        match value {
-            '\'' => Ok(QuoteStyle::Single),
-            '"' => Ok(QuoteStyle::Double),
-            _ => Err(()),
-        }
-    }
+    Preserve,
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/options.rs:210]: We need to update the documentation of the `FormatOption::quote_style` field to make the change reflect on our website:

https://github.com/astral-sh/ruff/blob/578ddf1bb1e8edef50cda9981744ba5fe825867d/crates/ruff_workspace/src/options.rs#L2804-L2829

**Review by MichaReiser — COMMENTED** [review]
I made a few changes to the code and introduced a `QuoteChar` to avoid having to handle `Preserve` in situations where `Preserve` isn't a valid variant (and there's no good way of e.g. implementing `invert`). 

I also extended your tests by a few docstring examples and noticed that single-quoted docstrings remain unchanged (see failing tests). This isn't specific to `preserve`, the docstring quotes also remain unchanged when using `quote-style: single`. 

@charliermarsh I belive you changed the formatter to enforce double quotes for all docstrings and triple quoted strings. Was it an intentional decision to handle single quoted docstrings the same as inline strings or do you think this should be changed? 

Implementation note: The string formatting should know that it is a docstring by check its layout.

**Comment by MichaReiser** [comment]
I rebased the changes, updated the documentation, and gated the new option behind the `--preview` flag to get more feedback. 

Something I noticed when updating the documentation is that `preserve` is special in the sense that it won't pick the quote that requires the few escapes, at least not the way it's implemented now. This is a slight deviation from `single` and `double` but I think it's subtle enough that adding `preserve` to quote-style is fine.

The reason for gating this behind preview is that we had some internal discussions and @charliermarsh brought up this valid concern:

> I think more users are likely to be confused that preserve changes docstrings than they are happy that preserve changes docstrings.

We gate this behind preview because we want to ship the functionality that allows users to preserve their quote style, but we need to gather more feedback if configuring this using `quote-style` is the right way to go. 


@charliermarsh or @BurntSushi would appreciate if you could review the changes that I made to this PR.

**Review by MichaReiser — APPROVED** [review]

**Review by BurntSushi — APPROVED** [review]
Nice!

One other thought here is whether it makes sense to document the nooks and crannies of the `preserve` option. That is, its impact (or not) on docstring quoting and how it interacts with escaping and what not.

**Review by charliermarsh — APPROVED** [review]

File: crates/ruff_python_formatter/tests/snapshots/format@quote_style.py.snap [hunk 2]
```diff
@@ -204,6 +248,22 @@ rb"""br double triple"""
...
+def docstring_double_triple():
+    """double triple"""
+
+
+def docstring_double():
+    "double triple"
+
+
+def docstring_single():
+    "single"
```
[the remark this answers is not in the document]
Comment (BurntSushi) [reply crates/ruff_python_formatter/tests/snapshots/format@quote_style.py.snap:266]: SGTM!

**Comment by MichaReiser** [comment]
> I agree with this, although like I mentioned this can be relaxed in a followup without breaking compatibility.

I continue to feel conflicted about this. The documentation for `quote-style` already become more complicated because some parts no longer apply for `preserve`. Potentially preserving all quotes styles and having this configured under the same setting will make this worse. 

The other challenge is that we wouldn't be able to support enforcing double quotes for triple quoted strings and docstrings but preserve the other quotes. 

Anyway, I think this is something we can re-visit when we get more feedback. This looks good to me. Thanks for pushing us to revisit adding this option by creating the PR.

**Comment by genericmoniker** [comment]
I was trying out the "preserve" quote-style on a code base formatted with black using "skip-string-normalization" and was surprised to see non-docstring multiline strings using `'''` have their quotes replaced. I came looking for a bug on that, but discovered it was intentional.

> enforces double quotes for triple-quoted strings (docstrings and multiline strings)

Just a little feedback.

**Comment by MichaReiser** [comment]
Thanks for the feedback. Does this behavior prevent you from adopting ruff or are okay with the changes that it introduces?

**Comment by genericmoniker** [comment]
I wouldn't say the behavior prevents adoption, but it does add a little friction. If there were only a small number of changes between black and ruff on our code base, I'd just open a pull request to switch to ruff and expect my team to be happy that formatting is faster. Without "preserve" there are hundreds of files changed. Since "preserve" is only currently available with the preview setting and that changes other things, using it also results in many files reformatted. 

I'm personally fine normalizing all quotes, and I imagine my teammates probably are too, but it becomes a point of discussion when lots of files change and cause merge conflicts with their outstanding pull requests.

**Comment by lucas-labs** [comment]
Just like @genericmoniker I also expected `preserve` to completely skip quotes normalization (just like black's `--skip-string-normalization` which is what we've been using on our code base). I understand this is by-default but I guess it would at least delay adopting ruff until we have fewer ongoing pull requests to avoid conflicts (I mean, to adopt it as a formatter, obviously; we are already using ruff as a linter and its awesome).

**Pull request merged**